packages/core/src/external/carequality/dr/get-document-download-url.ts

```diff
@@ -17,48 +17,50 @@ export async function buildDocumentReferences(
 ): Promise<DocumentReference[]> {
   validateBasePayload(payload);
 
-  const documentIds = extractDocumentIds(payload);
+  const [documentIds, uniqueIds] = extractDocumentIds(payload);
   if (documentIds.length === 0) {
     throw new XDSRegistryError("Valid Document ID is not defined");
   }
 
-  return await retrieveDocumentReferences(documentIds);
+  return await retrieveDocumentReferences(documentIds, uniqueIds);
 }
 
-async function retrieveDocumentReferences(documentIds: string[]): Promise<DocumentReference[]> {
+async function retrieveDocumentReferences(
+  documentIds: string[],
+  uniqueIds: string[]
+): Promise<DocumentReference[]> {
   const s3Utils = new S3Utils(region);
-  const documentReferences: DocumentReference[] = [];
-
-  // TODO consider making this more robust, so if one fails we still return the rest
-  for (const id of documentIds) {
-    const { size, contentType, eTag } = await s3Utils.getFileInfoFromS3(
-      id,
-      medicalDocumentsBucketName
-    );
-    if (!eTag) {
-      const message = `Failed to retrieve ETag for document`;
+  const documentReferencesPromises = documentIds.map(async (id, index) => {
+    const { size, contentType } = await s3Utils.getFileInfoFromS3(id, medicalDocumentsBucketName);
+    const uniqueId = uniqueIds[index];
+    if (!uniqueId) {
+      const message = `Failed to retrieve uniqueId for document`;
       console.log(`${message}: ${id}`);
       throw new XDSRegistryError("Failed to retrieve Document");
     }
-    const documentReference: DocumentReference = {
+    return {
       homeCommunityId: METRIPORT_HOME_COMMUNITY_ID,
       repositoryUniqueId: METRIPORT_REPOSITORY_UNIQUE_ID,
-      docUniqueId: eTag,
+      docUniqueId: uniqueId,
       contentType: contentType,
       size: size,
       urn: id,
     };
-    documentReferences.push(documentReference);
-  }
-
-  return documentReferences;
+  });
+  const documentReferences = await Promise.allSettled(documentReferencesPromises);
+  const successfulDocRefs = documentReferences.flatMap(p =>
+    p.status === "fulfilled" ? p.value : []
+  );
+  return successfulDocRefs;
 }
 
-function extractDocumentIds(payload: InboundDocumentRetrievalReq): string[] {
+function extractDocumentIds(payload: InboundDocumentRetrievalReq): [string[], string[]] {
```

Comment:
nit: Minor thing but since this function gets called before `retrieveDocumentReferences` it should be above it. Makes it easier to navigate the file

Reply:
I'd rather have smaller functions below... so one can read the code from top to bottom - in addition to see the more important things first. 😃

Reply:
Adding it to Notion so we discuss when we review the code standards.